polkadot/runtime/parachains/src/paras_inherent/mod.rs

```diff
@@ -529,11 +529,32 @@ impl<T: Config> Pallet<T> {
 
 		// Process new availability bitfields, yielding any availability cores whose
 		// work has now concluded.
-		let freed_concluded =
+		let (enact_weight, freed_concluded) =
 			inclusion::Pallet::<T>::update_pending_availability_and_get_freed_cores(
 				&validator_public[..],
 				bitfields.clone(),
 			);
+		all_weight_after.saturating_accrue(enact_weight);
+		log::debug!(
+			target: LOG_TARGET,
+			"Enacting weight: {}, all weight: {}",
+			enact_weight.ref_time(),
+			all_weight_after.ref_time(),
+		);
+
+		// It's possible that that after the enacting the candidates, the total weight
+		// goes over the limit, however, we can't do anything about it at this point.
+		// By using the `Mandatory` weight, we ensure the block is still accepted,
+		// but no other (user) transactions can be included.
+		if all_weight_after.any_gt(max_block_weight) {
```

Comment:
How are we using this runtime-computed weight?
I see this comment on the `PostDispatchInfo`:

```
/// NOTE: This can only be used to reduce the weight or fee, not increase it.
```

Where is it used to not allow further extrinsics?

Reply:
Here's my limited understanding (so please double check my thinking) of how weight limiting would work:

When we're building a block, we first apply inherents:
https://github.com/paritytech/polkadot-sdk/blob/149c70938f2b29f8d92ba1cc952aeb63d4084e27/substrate/client/basic-authorship/src/basic_authorship.rs#L340
`ParasInherentData` weight will be noted by `CheckWeight` extention:
https://github.com/paritytech/polkadot-sdk/blob/149c70938f2b29f8d92ba1cc952aeb63d4084e27/substrate/frame/system/src/extensions/check_weight.rs#L106-L113
and after that try to apply regular transactions (extrinsics):
https://github.com/paritytech/polkadot-sdk/blob/149c70938f2b29f8d92ba1cc952aeb63d4084e27/substrate/client/basic-authorship/src/basic_authorship.rs#L343-L347


If we have already reached the total block weight limit, no further extrinsics could be included since pre_dispatch would return an `ExhaustsResources` error:
https://github.com/paritytech/polkadot-sdk/blob/149c70938f2b29f8d92ba1cc952aeb63d4084e27/substrate/frame/system/src/extensions/check_weight.rs#L182-L193

And as such, no further transactions could be included in the block:
https://github.com/paritytech/polkadot-sdk/blob/149c70938f2b29f8d92ba1cc952aeb63d4084e27/substrate/client/basic-authorship/src/basic_authorship.rs#L502-L521

Reply:
This sounds about right to me. However, I'm not sure where we update the actual consumed weight.
The code you linked here https://github.com/paritytech/polkadot-sdk/blob/149c70938f2b29f8d92ba1cc952aeb63d4084e27/substrate/frame/system/src/extensions/check_weight.rs#L106-L113

is part of the pre-dispatch. 

I don't see where the post dispatch weight info is used to increase the weight after the inherent is applied.
AFAICT it's only being used to decrease it if it is lower than the static weight:
https://github.com/paritytech/polkadot-sdk/blob/0b52a2c19ebcf5d7a0d07974b70aec656704d249/substrate/frame/system/src/extensions/check_weight.rs#L254-L265

Reply:
I think the weight can be increased by using `register_extra_weight_unchecked`:
https://github.com/paritytech/polkadot-sdk/blob/0b52a2c19ebcf5d7a0d07974b70aec656704d249/substrate/frame/system/src/lib.rs#L1767

Otherwise, I think the overall block weight is increased in https://github.com/paritytech/polkadot-sdk/blob/0b52a2c19ebcf5d7a0d07974b70aec656704d249/substrate/frame/system/src/lib.rs#L2022, which also takes the minimum between the static weight and the measured weight (so it cannot increase it)

Reply:
Ok, now I see your point @alindima. Thanks for bringing this up! Indeed, it looks like the post dispatch weight can only go down.

What we could do is add the cost of enact_candidate * n_cores to the static/pre_dispatch `paras_inherent_total_weight` assuming some worst case messages number.

Another possiblity: just had a call with @kianenigma and we discussed that it might be a good idea to simply set the pre dispatch weight for para inherent to a (mandatory) max weight.